### PR TITLE
Add Bidirectional RNN

### PR DESCRIPTION
**Pull request by saksham189**
TODOS

- [x] Add test

Refer https://github.com/mlpack/mlpack/issues/1580

### CODE REVIEW

**Comment by saksham189** [comment]
I am not too sure about the merge operation of the two RNNs. Currently I concatenate the outputs and pass that through `LogSoftmax` layer. But I think that we might need to have a fully connected layer before that for the bidirectional RNN to be useful. 
Should we use FFN to perform the merge operation? @zoq

**Comment by saksham189** [comment]
I can't find any tests specific for bidirectional RNN. I have used one of the tests for RNN to test my implementation.

**Comment by zoq** [comment]
If I remember right Alex Graves used the Reber Grammar test (already implemented) for BLSTM as well.

**Comment by zoq** [comment]
About the merge operation, we could directly merge the results inside the BRNN class; let me know what you think.

> Currently I concatenate the outputs and pass that through LogSoftmax layer

Shouldn't we take the mean over the two?

**Comment by saksham189** [comment]
> Shouldn't we take the mean over the two?

There are multiple ways in which the outputs can be combined ie. summing, average and concatenating. Concatenate is used as the default in most places. Then we might apply a softmax layer or other fully connected layers to produce the final output.

**Comment by zoq** [comment]
Do you think we could introduce another template parameter something like: ``typename MergeLayerType = Concat<>()``?

**Comment by saksham189** [comment]
I would like to do that but the concat layer does a horizontal concatenation as compared to a vertical. 
https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/concat_impl.hpp#L64

Should I add a parameter to the layer?

**Comment by saksham189** [comment]
Also, I think that we would be required to add a `run` parameter to `Concat` as in `AddMerge`. 
I think in the forward pass of the concat we want `run` to be set to false since `input` to each layer inside concat would be different and set to true when calling Backward. So, we should create a visitor to set value of `run` parameter.

**Comment by zoq** [comment]
> Should I add a parameter to the layer or modify the current implementation? @zoq 

By modify the current implementation you mean the concat layer? Might be the best option here.

> Also, I think that we would be required to add a run parameter to Concat as in AddMerge.

Agreed, do you like to do this as part of the PR?

**Comment by saksham189** [comment]
@zoq this is ready for review. 

> If I remember right Alex Graves used the Reber Grammar test (already implemented) for BLSTM as well.

The paper does not have any results on BLSTM over reber grammar test. But I was able to find that BLSTMs can converge over reber grammar test in 1000 iterations as compared to LSTMs which take 1000 iterations and so, I have implemented a test on that.

Also, I have followed the BRNN algorithm described in Alex Graves paper for implementing the brnn class.
screenshot 2019-01-20 at 7 29 43 pm
screenshot 2019-01-20 at 7 29 49 pm

File: src/mlpack/methods/ann/visitor/run_set_visitor_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/**
+ * @file run_set_visitor_impl.hpp
+ * @author Marcus Edel
```
Comment (zoq) [inline src/mlpack/methods/ann/visitor/run_set_visitor_impl.hpp]: Since you wrote the code, you should adjust the name here :)

File: src/mlpack/tests/recurrent_network_test.cpp [hunk 2]
```diff
@@ -581,6 +653,125 @@ void ReberGrammarTestNetwork(const size_t hiddenSize = 4,
   BOOST_REQUIRE_GE(successes, 1);
 }
 
+/**
+ * Train the specified BRNN network and the construct a Reber grammar dataset.
+ */
+template<typename RecurrentLayerType>
+void ReberGrammarTestNetworkBRNN(const size_t hiddenSize = 4,
```
Comment (zoq) [inline src/mlpack/tests/recurrent_network_test.cpp]: Do you think we could pass a model instantiation to make the test code more general? In this case we could we don't have to write a special version for each network. Let me know what you think.

Comment (saksham189) [reply src/mlpack/tests/recurrent_network_test.cpp]: Yes that does reduce quite a bit code duplication. Although I had to add one if condition as well since the layers in the model were different.

File: src/mlpack/methods/ann/brnn_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+    forwardGradient.zeros();
+    for (size_t l = 0; l < networkSize; ++l)
+    {
+      boost::apply_visitor(LoadOutputParameterVisitor(
+          std::move(forwardRNNOutputParameter)),
+          forwardRNN.network[networkSize - 1 - l]);
+    }
+    boost::apply_visitor(BackwardVisitor(std::move(
+        boost::apply_visitor(outputParameterVisitor, forwardRNN.network.back())),
+        std::move(allDelta[rho - seqNum - 1]), std::move(delta)), mergeLayer);
```
Comment (saksham189) [inline src/mlpack/methods/ann/brnn_impl.hpp]: I call `Backward()` here on `mergeLayer` passing input from `forwardRNN` and similarly I have to do this while passing the input from `backwardRNN` (because mergeLayer can only take 1 input). This would call backward for each layer inside mergeLayer with the given input. 
Do you think we could add a parameter to `Backward()` inside mergeLayer so that it calls `Backward()` on a specific layer inside the merge layers? @zoq

Comment (zoq) [reply src/mlpack/methods/ann/brnn_impl.hpp]: Or provide another ``Backward`` overload that takes another input?

Comment (saksham189) [reply src/mlpack/methods/ann/brnn_impl.hpp]: Yes I have added the overload. Please let me know what you think.

Comment (zoq) [reply src/mlpack/methods/ann/brnn_impl.hpp]: Looks good to me 👍 ; will take another look at the brrn class tomorrow.

Comment (saksham189) [reply src/mlpack/methods/ann/brnn_impl.hpp]: okay thanks!

**Comment by saksham189** [comment]
Hey @zoq ,could you do a quick review.

File: src/mlpack/methods/ann/layer/concat_impl.hpp [hunk 4]
```diff
@@ -86,62 +69,70 @@ template<typename eT>
...
     arma::mat delta;
-    if (gy.n_cols == 1)
-    {
-      delta = gy.submat(j, 0, j + elements - 1, 0);
-    }
-    else
+    for (size_t i = 0; i < network.size(); ++i)
     {
-      delta = gy.submat(0, i, elements - 1, i);
+      size_t rows = boost::apply_visitor(outputParameterVisitor, network[i]).n_rows;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/concat_impl.hpp]: Do you think we should add a comment here, to provide some insight what this code does?

File: src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp [hunk 5]
```diff
@@ -21,20 +21,48 @@ namespace ann {
...
 {
   layer->Backward(std::move(input), std::move(error), std::move(delta));
 }
 
+template<typename T>
+inline typename std::enable_if<
+    HasRunCheck<T, bool&(T::*)(void)>::value, void>::type
+BackwardVisitor::LayerBackward(T* layer, arma::mat& /* input */) const
+{
+  if (index  == -1)
```
Comment (zoq) [inline src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp]: We compare against an unsigned integer (``size_t``) so this is never true.

File: src/mlpack/methods/ann/layer/concat_impl.hpp (hunk 4 above)
Comment (saksham189) [reply src/mlpack/methods/ann/layer/concat_impl.hpp]: Yes, would be better.

File: src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp (hunk 5 above)
Comment (saksham189) [reply src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp]: i have changed to int now

**Comment by zoq** [comment]
@saksham189 Just a note, if you update the PR, I think it's useful that the changes you made are visible, this makes the review a lot easier since the reviewer can see right away what has changed (between updates). Let me know what you think.

**Comment by saksham189** [comment]
> @saksham189 Just a note, if you update the PR, I think it's useful that the changes you made are visible, this makes the review a lot easier since the reviewer can see right away what has changed (between updates). Let me know what you think.

Yup. sorry about the inconvenience. Will push changes in new commits from now on.

**Comment by zoq** [comment]
@mlpack-jenkins test this please

**Comment by saksham189** [comment]
@zoq this is ready for review now.

File: src/mlpack/methods/ann/visitor/gradient_visitor.hpp [hunk 6]
```diff
@@ -30,7 +30,7 @@ class GradientVisitor : public boost::static_visitor<void>
  public:
   //! Executes the Gradient() method of the given module using the input and
   //! delta parameter.
-  GradientVisitor(arma::mat&& input, arma::mat&& delta);
+  GradientVisitor(arma::mat&& input, arma::mat&& delta, int layer = -1);
```
Comment (zoq) [inline src/mlpack/methods/ann/visitor/gradient_visitor.hpp]: Do you think instead of using ``int layer = -1`` we could provide a second constructor:

```c++
GradientVisitor(arma::mat&& input, arma::mat&& delta) : hasIndex(false), index(0)
{
}

GradientVisitor(arma::mat&& input, arma::mat&& delta, const size_t index) : hasIndex(true), index(index)
{
}
```

Let me know what you think.

Comment (saksham189) [reply src/mlpack/methods/ann/visitor/gradient_visitor.hpp]: alright sounds good to me

Comment (saksham189) [reply src/mlpack/methods/ann/visitor/gradient_visitor.hpp]: @zoq i have made the changes

**Comment by ShikharJ** [comment]
Nice work @saksham189. I'll review in the next few days and then we can merge this in.

File: src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp [hunk 7]
```diff
@@ -24,17 +24,58 @@ inline BackwardVisitor::BackwardVisitor(arma::mat&& input,
...
+  {
+    layer->Backward(std::move(input), std::move(error),
+        std::move(delta));
+  }
+  else
+  {
+    layer->Backward(std::move(input), std::move(error),
+        std::move(delta), index);
+  }
+
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/visitor/backward_visitor_impl.hpp]: Extra newline here.

File: src/mlpack/methods/ann/visitor/gradient_visitor_impl.hpp [hunk 8]
```diff
@@ -34,13 +46,33 @@ inline void GradientVisitor::operator()(LayerType* layer) const
...
+  {
+    layer->Gradient(std::move(input), std::move(delta),
+        std::move(layer->Gradient()));
+  }
+  else
+  {
+    layer->Gradient(std::move(input), std::move(delta),
+        std::move(layer->Gradient()), index);
+  }
+
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/visitor/gradient_visitor_impl.hpp]: Same here.

**Comment by zoq** [comment]
One more note, if you do commit something, it would be helpful if the commit message does reflect the changes you did. That makes searching through the commit history a lot easier.

File: src/mlpack/methods/ann/brnn_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,671 @@
...
+    gradient.zeros();
+  }
+
+  if (backwardGradient.is_empty())
+  {
+    backwardGradient = arma::zeros<arma::mat>(
+        parameter.n_rows/2,
+        parameter.n_cols);
+    forwardGradient = arma::zeros<arma::mat>(
+        parameter.n_rows/2,
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/brnn_impl.hpp]: Can you add a space between the ` \ 2` here and above?

File: src/mlpack/methods/ann/brnn_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,671 @@
...
+void BRNN<OutputLayerType, MergeLayerType, MergeOutputType,
+    InitializationRuleType, CustomLayers...>::ResetDeterministic()
+{
+  forwardRNN.deterministic = this->deterministic;
+  backwardRNN.deterministic = this->deterministic;
+  forwardRNN.ResetDeterministic();
+  backwardRNN.ResetDeterministic();
+}
+
+
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/brnn_impl.hpp]: Extra newline used here.

**Comment by saksham189** [comment]
> One more note, if you do commit something, it would be helpful if the commit message does reflect the changes you did. That makes searching through the commit history a lot easier.

I will remember to follow this from now onwards. I have changed the previous commit message as well. @zoq

**Comment by zoq** [comment]
Great, @saksham189 do you like to add anything else here, or is this ready?

**Comment by saksham189** [comment]
This is ready from my side. Unless you have anything to add. @zoq

**Comment by zoq** [comment]
Thanks again for the great contribution.
